src/rpc/WorkQueue.cpp

```diff
@@ -69,25 +81,145 @@ WorkQueue::WorkQueue(std::uint32_t numWorkers, uint32_t maxSize)
           "The current number of tasks in the queue"
       )}
     , ioc_{numWorkers}
+    , strand_{ioc_.get_executor()}
+    , waitTimer_(ioc_)
 {
     if (maxSize != 0)
         maxSize_ = maxSize;
+
+    util::spawn(strand_, [this](auto yield) { dispatcherLoop(yield); });
 }
 
 WorkQueue::~WorkQueue()
 {
-    join();
+    stop();
+}
+
+bool
+WorkQueue::postCoro(std::function<void(boost::asio::yield_context)> func, bool isWhiteListed, Priority priority)
+{
+    if (stopping_) {
+        LOG(log_.warn()) << "Queue is stopping, rejecting incoming task.";
+        return false;
+    }
+
+    if (size() >= maxSize_ && !isWhiteListed) {
+        LOG(log_.warn()) << "Queue is full. rejecting job. current size = " << size() << "; max size = " << maxSize_;
+        return false;
+    }
+
+    ++curSize_.get();
+    auto needsWakeup = false;
+
+    {
+        auto state = dispatcherState_.lock();
+
+        if (state->isIdle) {
+            needsWakeup = true;
+            state->isIdle = false;
+        }
+
+        state->push(priority, std::move(func));
+    }
+
+    if (needsWakeup)
+        boost::asio::post(strand_, [this] { waitTimer_.cancel(); });
+
+    return true;
 }
 
 void
-WorkQueue::stop(std::function<void()> onQueueEmpty)
+WorkQueue::dispatcherLoop(boost::asio::yield_context yield)
+{
+    LOG(log_.debug()) << "WorkQueue dispatcher starting";
```

Comment:
maybe info here?

Reply:
I have no strong opinion on this so i'll set back to info